src/components/Form.js

```diff
@@ -0,0 +1,220 @@
+
+        /** Server side error message */
+        serverErrorMessage: PropTypes.string,
+    }),
+
+    /** Contains draft values for each input in the form */
+    // eslint-disable-next-line react/forbid-prop-types
+    draftValues: PropTypes.object,
+
+    ...withLocalizePropTypes,
+};
+
+const defaultProps = {
+    formState: {
+        isSubmitting: false,
+        serverErrorMessage: '',
+    },
+    draftValues: {},
+};
+
+class Form extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            errors: {},
+        };
+
+        this.inputRefs = {};
+        this.touchedInputs = {};
+
+        this.getValues = this.getValues.bind(this);
+        this.setTouchedInput = this.setTouchedInput.bind(this);
+        this.validate = this.validate.bind(this);
+        this.submit = this.submit.bind(this);
+    }
+
+    /**
+     * @param {String} inputID - The inputID of the input being touched
+     */
+    setTouchedInput(inputID) {
+        this.touchedInputs[inputID] = true;
+    }
+
+    /**
+     * @returns {Object} - An object containing the values for each inputID
+     */
+    getValues() {
+        const values = {};
+        _.each(this.inputRefs, (inputRef, inputID) => {
+            values[inputID] = inputRef.value;
+        });
+        return values;
+    }
+
+    submit() {
+        // Return early if the form is already submitting to avoid duplicate submission
+        if (this.props.formState.isSubmitting) {
+            return;
+        }
+
+        // Touches all form inputs so we can validate the entire form
+        _.each(this.inputRefs, (inputRef, inputID) => (
+            this.touchedInputs[inputID] = true
+        ));
+
+        // Validate form and return early if any errors are found
+        const values = this.getValues();
+        if (!_.isEmpty(this.validate(values))) {
+            return;
+        }
+
+        // Set loading state and call submit handler
+        FormActions.setIsSubmitting(this.props.formID, true);
+        this.props.onSubmit(values);
+    }
+
+    /**
+     * @param {Object} values - An object containing the value of each inputID, e.g. {inputID1: value1, inputID2: value2}
+     * @returns {Object} - An object containing the errors for each inputID, e.g. {inputID1: error1, inputID2: error2}
+     */
+    validate(values) {
+        FormActions.setServerErrorMessage(this.props.formID, '');
+        const validationErrors = this.props.validate(values);
+
+        if (!_.isObject(validationErrors)) {
+            throw new Error('Validate callback must return an empty object or an object with shape {inputID: error}');
+        }
+
+        const errors = _.pick(validationErrors, (inputValue, inputID) => (
+            Boolean(this.touchedInputs[inputID])
+        ));
+        this.setState({errors});
+        return errors;
+    }
+
+    childrenWrapperWithProps(children) {
+        // eslint-disable-next-line rulesdir/prefer-underscore-method
```

Comment:
We should be able to remove this now, right?

Reply:
Done!